### ISSUE TITLE
Expired access token not refreshed

### ISSUE THREAD

**Issue by hannes-ma** [BitzArt/Blazor.Auth#24]
Hello all,
I'm encountering a strange problem using Blazor.Auth on a Blazor Server implementation where the access token is not refreshed after expiration.

I have the following situation:
1. User logs in and gets the access + refresh token. The access token is valid 2 minutes while the refresh token is valid 2 months.
2. User stays on one page without doing any operation until the access token expires (after 2 minutes)
3. User navigates to another page. Sometimes the ServerSidePrerenderAuthenticationStateProvider, which handles the access/refresh token, still sees the access token cookie in the HTTPRequest and passes it on to BuildClaimsPrincipal(). Inside that function the validation of the token then obviously fails as the access token is expired and throws an exception which results in an error to the user instead of attempting to refresh the token.

Just an assumption, but could it be that the server and the browser have a slightly different time of day, so the cookie hasn't been expired and removed on the browser so it is still sending it to the server? I've seen this using Google Chrome.

What could be a possible fix around this? IMO the code in ServerSidePrerenderAuthenticationStateProvider should be changed to allow the access token to fail validation and in this case it should go ahead with the refresh token.

**Comment by YuriyDurov** [comment]
Hello @hannes-ma 

Thank you for pointing this out, we seem to have been experiencing the same problem. Please let us know if you discover any more details about this issue. We will try to investigate in the near future.

**Comment by jstafford5380** [comment]
Is this an issue with mix mode where blazor switching back and forth between client components and server components? There were some Auth0 articles that were talking about how complex that actually is and it was causing similar symptoms.

**Comment by hannes-ma** [comment]
@jstafford5380 
Well in my case I'm only using server-side rendering (no WASM and also pre-rendering disabled). What I've seen till now is that the issue is not reproducible on my local development environment, where the blazor application and browser are running on the same host. It only happens when deployed via docker container on the cloud server, therefore my assumption that it depends on date/time differences between server and client browser.

**Comment by YuriyDurov** [comment]
@hannes-ma 

Agreed, the issue does not seem to relate to the complexity of `InteractiveAuto`, as we are experiencing it while using just `InteractiveWebAssembly` rendering.

Probably something to do with timing and at what exact moment expired cookies are getting cleared from the browser. We have implemented the package assuming the expired cookies will no longer be sent after they have expired, but there is clearly something going wrong there.

**Comment by YuriyDurov** [comment]
At first it seemed like we were experiencing the same issue as the one described here, but at some point it went away. Not sure what caused it to change. I have also tried manually to get it to happen but wasn't able to.

Is anyone here able to replicate this issue at least somewhat consistently? We seem to require additional information on this issue in order to track it down.

**Comment by YuriyDurov** [comment]
We have never exactly figured out the root cause of this issue, but the library has received a major rework in the latest update. The part of the library expressing this behavior was almost fully rewritten and the underlying behaviors were changed drastically. The wasm side of the application now fully relies on the server part to resolve the user's `AuthenticationState`. Considering the aforementioned points, I find it very unlikely that we will face this same issue any more after this update.
